server/etcdserver/txn/txn.go

```diff
@@ -404,23 +373,36 @@ func executeTxn(ctx context.Context, lg *zap.Logger, txnWrite mvcc.TxnWrite, rt
 	return txns, nil
 }
 
-func checkPut(rv mvcc.ReadView, lessor lease.Lessor, req *pb.PutRequest) error {
-	if req.IgnoreValue || req.IgnoreLease {
-		// expects previous key-value, error if not exist
-		rr, err := rv.Range(context.TODO(), req.Key, nil, mvcc.RangeOptions{})
+func checkPut(rv mvcc.ReadView, lessor lease.Lessor, req *pb.PutRequest) (val []byte, prevKV *mvccpb.KeyValue, leaseID lease.LeaseID, err error) {
+	val, leaseID = req.Value, lease.LeaseID(req.Lease)
+
+	if req.IgnoreValue || req.IgnoreLease || req.PrevKv {
+		resp, err := rv.Range(context.TODO(), req.Key, nil, mvcc.RangeOptions{})
 		if err != nil {
-			return err
+			return nil, nil, 0, err
 		}
-		if rr == nil || len(rr.KVs) == 0 {
-			return errors.ErrKeyNotFound
+		if resp != nil && len(resp.KVs) != 0 {
+			prevKV = &resp.KVs[0]
 		}
+		if (req.IgnoreValue || req.IgnoreLease) && prevKV == nil {
+			// ignore_{lease,value} flag expects previous key-value pair
+			return nil, nil, 0, errors.ErrKeyNotFound
+		}
+	}
+	if req.IgnoreValue {
+		val = prevKV.Value
 	}
-	if lease.LeaseID(req.Lease) != lease.NoLease {
-		if l := lessor.Lookup(lease.LeaseID(req.Lease)); l == nil {
-			return lease.ErrLeaseNotFound
+	if req.IgnoreLease {
+		leaseID = lease.LeaseID(prevKV.Lease)
+	} else if leaseID != lease.NoLease {
+		if l := lessor.Lookup(leaseID); l == nil {
+			return nil, nil, 0, lease.ErrLeaseNotFound
```

Comment:
The following logic shouldn't be part of `checkPut`.

```
	if req.IgnoreValue {
		val = prevKV.Value
	}
	if req.IgnoreLease {
		leaseID = lease.LeaseID(prevKV.Lease)
	} 
	if !req.PrevKv {
		return val, nil, leaseID, nil
	}
```

Reply:
I can implement it that way, but it results in weird duplication of logic to check whether prevKV is needed.